### PR TITLE
nspawn: resolve network interface names before move to container namespace

### LINKED ISSUE #31104: `systemd-nspawn` fails to resolve altname with `--network-interface=` on older kernels
This seems to be fallout from https://github.com/systemd/systemd/pull/30956, but CI didn't catch it because Ubuntu CI was gone and the C8S job runs QEMU tests only once per day (where it failed, but I got to it just now). It can be easily reproduced with:

```
# ip link add dev sd-test type dummy && ip link property add dev sd-test altname foo; ip link show sd-test
6: sd-test: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether be:0e:8b:21:c8:ca brd ff:ff:ff:ff:ff:ff
    altname foo
# systemd-nspawn -D / --volatile=yes --network-interface=sd-test:foo true
░ Spawning container centos8s.localdomain on /.
░ Press Ctrl-] three times within 1s to kill container.

Unknown interface name foo: No such device
(sd-netns) failed with exit status 1.
Failed to fork process (sd-netns): Protocol error
```

It's the `rtnl_resolve_link_alternative_name()` call from `sd_device_new_from_ifname()` that fails in this case, but that's as far as I've gotten. /cc @yuwata

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #31104.

### CODE REVIEW

**Comment by mrc0mmand** [comment]
It does seem to fix the original issue, but stuff unfortunately breaks a couple of tests later:

```
Jan 27 22:48:04 H testsuite-13.sh[2634]: + for container in nss-mymachines-{noip,singleip,manyips}
Jan 27 22:48:04 H testsuite-13.sh[2634]: + timeout 30 bash -xec 'while [[ ! -e /var/lib/machines/nss-mymachines-manyips/initialized ]]; do sleep .5; done'
Jan 27 22:48:04 H testsuite-13.sh[2726]: + [[ ! -e /var/lib/machines/nss-mymachines-manyips/initialized ]]
Jan 27 22:48:04 H testsuite-13.sh[2634]: + ip addr add 10.1.0.1/24 dev ve-singleip
Jan 27 22:48:04 H testsuite-13.sh[2634]: + ip addr add 10.2.0.1/24 dev ve-manyips
Jan 27 22:48:04 H testsuite-13.sh[2634]: + ip addr add fd00:dead:beef:cafe::1/64 dev ve-manyips nodad
Jan 27 22:48:04 H testsuite-13.sh[2729]: RTNETLINK answers: Permission denied
```

It can be easily reproduced with something like:
```
# systemd-nspawn -bD / --volatile=yes --network-veth &
[1] 141930
░ Spawning container centos8s.localdomain on /.
░ Press Ctrl-] three times within 1s to kill container.
Using default interface naming scheme 'v255'.
Network interface name 've-centos8s.localdomain' has been changed to 've-centos8sb2Mq' to fit length constraints.

[1]+  Stopped                 systemd-nspawn -bD / --volatile=yes --network-veth
# ip l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:81:31:6b brd ff:ff:ff:ff:ff:ff
    altname enp0s6
    altname ens6
17: ve-centos8sb2Mq@if2: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether e6:f1:00:48:86:9b brd ff:ff:ff:ff:ff:ff link-netnsid 0
    altname ve-centos8s.localdomain
# ip a add 10.0.0.1/24 dev ve-centos8sb2Mq
# ip a add fd00:dead:beef:cafe::1/64 dev ve-centos8sb2Mq
RTNETLINK answers: Permission denied
```

**Comment by mrc0mmand** [comment]
Ah, that's not caused by this patch, since it fails with the same error even on main, the altname test just fails first. So I guess it got broken also by https://github.com/systemd/systemd/pull/30956.

**Comment by yuwata** [comment]
I have no idea. Please confirm that IPv6 is not disabled on the system?
At least, with kernel v6.6.9 works fine with the above reproducer. Please bisect the kernel and find which commit fixes the issue.

**Comment by mrc0mmand** [comment]
> I have no idea. Please confirm that IPv6 is not disabled on the system? At least, with kernel v6.6.9 works fine with the above reproducer. Please bisect the kernel and find which commit fixes the issue. And, maybe better to move the new issue to a separated page. At least, the issue should not related to #30956 and this PR.

Ugh, sorry about that, I completely forgot about this snippet:
```
# grep ipv6 /etc/sysctl.conf 
net.ipv6.conf.all.disable_ipv6 = 1
```
(why netlink returns EPERM instead of EOPNOTSUPP???)

Anyway, with this sorted out, the test now works as expected. Thanks a lot!

**Review by mrc0mmand — APPROVED** [review]

**Comment by yuwata** [comment]
> (why netlink returns EPERM instead of EOPNOTSUPP???)

Yeah, but this is the behavior for long time.
https://github.com/torvalds/linux/blob/41bccc98fb7931d63d03f326a746ac4d429c1dd3/net/ipv6/addrconf.c#L2514-L2515

File: src/libsystemd/sd-netlink/netlink-util.h [hunk 1]
```diff
@@ -35,14 +51,30 @@ int rtnl_set_link_properties(
...
+}
+static inline int rtnl_resolve_ifname(sd_netlink **rtnl, const char *name) {
+        return rtnl_resolve_ifname_full(rtnl, RESOLVE_IFNAME_MAIN | RESOLVE_IFNAME_ALTERNATIVE, name, NULL, NULL);
+}
+static inline int rtnl_resolve_interface(sd_netlink **rtnl, const char *name) {
+        return rtnl_resolve_ifname_full(rtnl, _RESOLVE_IFNAME_ALL, name, NULL, NULL);
+}
+static inline int rtnl_resolve_interface_or_warn(sd_netlink **rtnl, const char *name) {
+        int r;
+
```
Comment (bluca) [inline src/libsystemd/sd-netlink/netlink-util.h:71]: assert(name) since it's used in the log, or change it to strna()

**Review by bluca — APPROVED** [review]

**Comment by mrc0mmand** [comment]
By the way, once this is merged, I'll go ahead and bump the C8S CentOS CI job to C9S (which, kernel-wise, means a bump from ~4.18.0 to ~5.14.0).

File: src/libsystemd/sd-netlink/netlink-util.c [hunk 2]
```diff
@@ -11,6 +11,156 @@
...
+
+                r = sd_netlink_call(*rtnl, message, 0, &reply);
+                if (r >= 0)
+                        return parse_newlink_message(reply, ret_name, ret_altnames);
+                /* The kernels older than 76c9ac0ee878f6693d398d3a95ccaf85e1f597a6 (v5.5) return -EINVAL. */
+                if (!IN_SET(r, -ENODEV, -EINVAL))
+                        return r;
+        }
+
+        /* Finally, assume the string is a decimal formatted ifindex. */
```
Comment (poettering) [inline src/libsystemd/sd-netlink/netlink-util.c:148]: why put this last? I'd really put this first, given it's the cheapest operation

**Comment by poettering** [comment]
lgtm

File: src/libsystemd/sd-netlink/netlink-util.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd/sd-netlink/netlink-util.c:148]: No reason, but all operations should have the same cost.

**Pull request merged**